### PR TITLE
Add action list feed feature for upcoming events

### PR DESCRIPTION
**Pull request by vinnyjth**
### What does this PR do, or why is it needed?

- Adds the ability to show upcoming events inside the `ActionListFeature`
- Adds a method to fetch "upcoming public" events, or "all events" + a warning if you're not using the Apollos plugin. 
- Adds a method to fetch an image for a given event. 

### What design trade-offs/decisions were made?

Will result in a few additional issues being created

- Need to implement "READ_EVENT" action.
- Related #1029 

Chose not to go overboard making the "UPCOMING_EVENT" algorithm hyper customizable. We can take a look at enhanced functionality in future PRs. 

### How do I test this PR?

1. Open the home feed, you should see an event listed in the "This week at Apollos" section.
2. Look at my pretty picture below of a home feed configured just to show a single upcoming event

Screen Shot 2019-09-16 at 3 40 01 PM

## TODO:

- [x] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [tag-issues-here]
- [x] No new warnings in tests, in storybook, and in-app
- [x] Upload GIF(s) of iOS and Android
- [x] Set a relevant reviewer

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

File: packages/apollos-data-connector-rock/src/events/data-source.js [hunk 1]
```diff
@@ -17,13 +18,45 @@ export default class Event extends RockApolloDataSource {
       .filter(`CampusId eq ${id}`)
       .get();
 
-  getName = async (id) => {
+  findRecent = () => {
```
Comment (redreceipt) [inline packages/apollos-data-connector-rock/src/events/data-source.js]:
```suggestion
  getRecent = (numEvents = 3, campusID = null) => {
```

File: packages/apollos-data-connector-rock/src/events/data-source.js [hunk 2]
```diff
@@ -17,13 +18,45 @@ export default class Event extends RockApolloDataSource {
...
+      );
+    } else {
+      request = this.request(
+        `Apollos/GetEventItemOccurencesByCalendarId?id=${1}`
+      );
+    }
+    return request
+      .expand('Schedule')
+      .orderBy('Schedule/EffectiveStartDate')
+      .filter('Schedule/EffectiveStartDate ne null');
```
Comment (redreceipt) [inline packages/apollos-data-connector-rock/src/events/data-source.js]:
```suggestion
      .filter('Schedule/EffectiveStartDate ne null');
      .filter(campusID ? `CampusId eq ${campusID}` : "")
      .top(numEvents)
      .get()
```

File: packages/apollos-data-connector-rock/src/features/data-source.js [hunk 3]
```diff
@@ -63,6 +65,26 @@ export default class Features extends RockApolloDataSource {
     };
   }
 
+  async upcomingEventsAlgorithm() {
+    const { Events } = this.context.dataSources;
+
+    // Get the first three persona items.
+    const events = await Events.findRecent()
```
Comment (redreceipt) [inline packages/apollos-data-connector-rock/src/features/data-source.js]:
```suggestion
    const events = await Events.getRecent()
    
```

File: packages/apollos-data-connector-rock/src/features/data-source.js [hunk 4]
```diff
@@ -63,6 +65,26 @@ export default class Features extends RockApolloDataSource {
...
+      .top(3)
```
Comment (redreceipt) [inline packages/apollos-data-connector-rock/src/features/data-source.js]:
```suggestion
```

File: packages/apollos-data-connector-rock/src/features/data-source.js [hunk 5]
```diff
@@ -63,6 +65,26 @@ export default class Features extends RockApolloDataSource {
...
+      .get();
```
Comment (redreceipt) [inline packages/apollos-data-connector-rock/src/features/data-source.js]:
```suggestion

```

**Comment by redreceipt** [comment]
This is awesome. It was a little confusing to me to have a datasource function that was a request builder so I made it a getter, little more clear in my opinion. Also, the `getByCampus` function is potentially dead code without considering the new endpoint so I extended the other getter. maybe you have some other ideas about future campus implementation?

**Comment by vinnyjth** [comment]
>This is awesome. It was a little confusing to me to have a datasource function that was a request builder so I made it a getter, little more clear in my opinion. 

I understand that returning a cursor rather than a set of data is confusing, but it's important if we ever want to use our pagination helpers or add additional filtering to a request. We do this a lot with ContentItems (pretty much every method on the data source returns a cursor, rather than a feed of content) 

I am open for changing the name of the method to better reflect that, though :) maybe something like `getRecentCursor`?

**Comment by redreceipt** [comment]
gotcha, what about `getRecentRequest`? I'm not making the connection to a cursor, isn't that a field on a `Connection` data type for pagination?

**Comment by redreceipt** [comment]
and this case, we can just extend the `getByCampus` function to first go get the request and then add the filter in the same way. I like it.

**Comment by vinnyjth** [comment]
A `cursor` is our word for a `request` that hasn't been fetched yet. It's pretty much any instance of the `request` class. 

Pagination can be done on any any instance of the `request` class. If you complete the request by calling `get` or `first`, you can't paginate. (Well, you can, you can't use your pagination helpers though)

**Comment by vinnyjth** [comment]
> and this case, we can just extend the getByCampus function to first go get the request and then add the filter in the same way. I like it.

Yep! That's one of the advantages. We do that a lot in `ContentItem`

**Comment by redreceipt** [comment]
just so I'm clear, pagination here and graphql pagination are two different things right? I think that's what's confusing me? same with cursors?

**Comment by vinnyjth** [comment]
What our helper does is allow us to implement graphql pagination easily. You are correct, cursor's are totally different.

**Comment by redreceipt** [comment]
cool. disregard my suggestions then :) rename anything that makes sense to you, if not no sweat. Let me test and give a thumbs up.

**Comment by vinnyjth** [comment]
Here's a good example of pagination: 

https://github.com/ApollosProject/apollos-prototype/blob/browser/packages/apollos-data-connector-rock/src/content-channels/resolver.js#L12

Because `dataSources.ContentItem.byContentChannelId(id)` is a cursor/request, in this example, we can implement the graphql pagination logic very easily using this helper

**Comment by redreceipt** [comment]
for the `subtitle` you're using moment to do something with the schedule but I've already got a function for the start and end date/time

**Comment by redreceipt** [comment]
I say that because it's wrong in the client as is

**Comment by vinnyjth** [comment]
> I say that because it's wrong in the client as is.

Good catch. Little strange that it's wrong, because I'Do you want refactor to use your method and see if it fixes it?

**Comment by redreceipt** [comment]
Yep just give me a bit, reviewing a PR for dan right now

**Comment by redreceipt** [comment]
actually this just made me realize I can improve that function too and give us a slight speed boost. stay tuned.

**Comment by redreceipt** [comment]
ok take a look at this

**Comment by vinnyjth** [comment]
Kicking butt and taking names @redreceipt !

**Comment by redreceipt** [comment]
this one is doing weird things with tests too now...

**Pull request merged**